lib/srv/db/common/auth.go

```diff
@@ -296,23 +296,36 @@ func (a *dbAuth) GetTLSConfig(ctx context.Context, sessionCtx *Session) (*tls.Co
 	tlsConfig := &tls.Config{
 		RootCAs: x509.NewCertPool(),
 	}
-	// Don't set the ServerName when connecting to a MongoDB cluster - in case
-	// of replica set the driver may dial multiple servers and will set
-	// ServerName itself. For Postgres/MySQL we're always connecting to the
-	// server specified in URI so set ServerName ourselves.
-	if sessionCtx.Database.GetProtocol() != defaults.ProtocolMongoDB {
+
+	dbTLSConfig := sessionCtx.Database.GetTLS()
+
+	if dbTLSConfig.ServerName != "" {
+		// Use user provided server name.
+		tlsConfig.ServerName = dbTLSConfig.ServerName
+	} else if sessionCtx.Database.GetProtocol() != defaults.ProtocolMongoDB {
+		// Don't set the ServerName when connecting to a MongoDB cluster - in case
```

Comment:
The comment says "Don't set the ServerName when connecting to a MongoDB cluster" but if ServerName was provided the `tlsConfig.ServerName` will be set also for Mongo because first if condition `dbTLSConfig.ServerName != ""`  will be evaluated to true.